### PR TITLE
Include Additional API Tests

### PR DESCRIPTION
**Pull request by DyrellC**
This PR includes the addition tests for: 
1. `getInclusionStates`
2. `wereAddressesSpentFrom`
3. `getBalances`
4. `getInclusionStates`
5. `attachToTangle`
6. `storeTransactions`
7. `findTransactions`
8. `interruptAttachToTangle`
9. `checkConsistency`

It also includes a fix for the `addNeighbors` and `removedNeighbors` calls, and a structure change for the dynamic api calling step.

### CODE REVIEW

File: PythonRegression/tests/features/steps/api_test_steps.py [hunk 1]
```diff
@@ -75,20 +91,38 @@ def spam_call_gtta(step,numTests,node):
...
         responseVal.append(response)
         
-    responses[apiCall] = {}
-    responses[apiCall][node] = responseVal
+    world.responses[apiCall] = {}
+    world.responses[apiCall][node] = responseVal
+
+
+@step(r'getInclusionStates is called with the transaction "([^"]*)" and tips "([^"]*)" on "([^"]*)"')
+def inclusion_state_is_called(step,transaction,tips,node):
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/api_test_steps.py]: Actually I am a little confused. The previous function seemed to be a general way to call *all* api methods. Why do you need also functions for specific calls.

Either have a function for specific calls or a general function.

Imo if you will adopt the gherkin table offer for the general function you will have to write less python if we add a new api call.

File: PythonRegression/tests/features/steps/api_test_steps.py [hunk 2]
```diff
@@ -97,7 +131,7 @@ def compare_response(step):
             assert str(responseKeys[response_key_val]) == str(keys[response_key_val]['keys']), "There was an error with the response" 
     
     elif apiCall == 'getNeighbors' or apiCall == 'getTips':
-        responseList = responses[apiCall][nodeId] 
+        responseList = world.responses[apiCall][nodeId]
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/api_test_steps.py]: IIUC the only difference between the 2 ifs is that in one case you return a response list, and in the other case there is a single response.

If you treat the single response as a list with size == 1 then you can remove the `if-elif` statement.

File: PythonRegression/tests/features/steps/blowball_tests_steps.py [hunk 3]
```diff
@@ -9,25 +9,25 @@
 
 tests = api_test_logic
 
-config = {}
-responses = {}
-test_vars = {}
+world.config = {}
+world.responses = {}
+world.test_vars = {}
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/blowball_tests_steps.py]: you again have a PR that does 2 things. Add API calls, and start using `world`... (refrain from dong this next time. Use separate PRs. It is supposedly fine that you started to use `world` in the api tests in this PR, but this file is out of scope)
If you start using world, you should have a hook that cleans it after each feature is complete.

**Comment by DyrellC** [comment]
@GalRogozinski ready for another review

File: PythonRegression/tests/features/machine1/1_api_tests.feature [hunk 4]
```diff
@@ -38,56 +53,174 @@ Feature: Test API calls on Machine 1
...
-		
+		|trunkTransaction					|
+
+
+    Scenario: CheckConsistency is called
+        Given "checkConsistency" is called on "nodeA" with:
+        |keys           |values             |type           |
+        |tails          |TEST_HASH          |staticList     |
+
+        Then a response with the following is returned:
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/machine1/1_api_tests.feature]: Since we have a static db, maybe you can check the returned values and not just the keys

File: PythonRegression/tests/features/machine1/1_api_tests.feature [hunk 5]
```diff
@@ -38,56 +53,174 @@ Feature: Test API calls on Machine 1
...
+
+
+    #Values can be found in util/static_vals.py
+    Scenario: GetInclusionStates is called
+	    Given "getInclusionStates" is called on "nodeA" with:
+        |keys           |values             |type               |
+        |transactions   |TEST_HASH          |staticList         |
+        |tips           |TEST_TIP_LIST      |staticValue        |
+
+        Then a response with the following is returned:
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/machine1/1_api_tests.feature]: Since we have a static db, maybe you can check the returned values and not just the keys

File: PythonRegression/tests/features/machine1/1_api_tests.feature [hunk 6]
```diff
@@ -38,56 +53,174 @@ Feature: Test API calls on Machine 1
...
+        |keys                               |
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/machine1/1_api_tests.feature]: Since we have a static db, maybe you can check the returned values and not just the keys

File: PythonRegression/tests/features/machine1/1_api_tests.feature (hunk 4 above)
Comment (GalRogozinski) [reply PythonRegression/tests/features/machine1/1_api_tests.feature]: For this test I will even find a case that returns `true` and a case that returns `false`.

File: PythonRegression/tests/features/machine1/1_api_tests.feature [hunk 7]
```diff
@@ -38,56 +53,174 @@ Feature: Test API calls on Machine 1
...
+
+    #Address can be found in util/static_vals.py
+    Scenario: GetBalances is called
+        Given "getBalances" is called on "nodeA" with:
+        |keys       |values                 |type               |
+        |addresses  |TEST_EMPTY_ADDRESS     |staticList         |
+        |threshold  |100                    |int                |
+
+
+        Then a response with the following is returned:
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/machine1/1_api_tests.feature]: Since we have a static db, maybe you can check the returned values and not just the keys

File: PythonRegression/tests/features/machine1/1_api_tests.feature [hunk 8]
```diff
@@ -38,56 +53,174 @@ Feature: Test API calls on Machine 1
...
+    Scenario: WereAddressesSpentFrom is called
+        Given "wereAddressesSpentFrom" is called on "nodeA" with:
+        |keys       |values                 |type               |
+        |addresses  |TEST_EMPTY_ADDRESS     |staticList         |
+
+    	Then a response with the following is returned:
+    	|keys                               |
+    	|duration                           |
+    	|states                             |
+
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/machine1/1_api_tests.feature]: Since we have a static db, maybe you can check the returned values and not just the keys

File: PythonRegression/tests/features/machine1/1_api_tests.feature [hunk 9]
```diff
@@ -38,56 +53,174 @@ Feature: Test API calls on Machine 1
...
+        that transaction via its address.
+
+        Given a transaction is generated and attached on "nodeA" with:
+        |keys       |values                 |type           |
+        |address    |TEST_STORE_ADDRESS     |staticValue    |
+        |value      |0                      |int            |
+
+        Then a response with the following is returned:
+        |keys                               |
+        |trytes                             |
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/machine1/1_api_tests.feature]: Since we have a static db, maybe you can check the returned values and not just the keys

File: PythonRegression/tests/features/machine1/1_api_tests.feature [hunk 10]
```diff
@@ -38,56 +53,174 @@ Feature: Test API calls on Machine 1
...
+        |keys       |values                 |type           |
+        |trytes     |TEST_STORE_TRANSACTION |staticValue    |
+
+        And "findTransactions" is called on "nodeA" with:
+        |keys       |values                 |type           |
+        |addresses  |TEST_STORE_ADDRESS     |staticList     |
+
+        Then a response with the following is returned:
+        |keys                               |
+        |hashes                             |
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/machine1/1_api_tests.feature]: Since we have a static db, maybe you can check the returned values and not just the keys

File: PythonRegression/tests/features/steps/api_test_steps.py [hunk 11]
```diff
@@ -13,52 +14,98 @@
 neighbors = static_vals.TEST_NEIGHBORS
 testAddress = static_vals.TEST_ADDRESS
 
+localVar = "TESTINGLOCALVARIABLESAAAAAAAAAAAAAAAAAAAAAAAAAAH"
+
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/api_test_steps.py]: Is it a local variable if it is global?
Is it a constant? maybe you want to write it as upper case as a reminder that it shouldn't change?

File: PythonRegression/tests/features/steps/api_test_steps.py [hunk 12]
```diff
@@ -13,52 +14,98 @@
...
 config = {}
 responses = {'getNodeInfo':{},'getNeighbors':{},'getTips':{},'getTransactionsToApprove': {},'getTrytes':{}}
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/api_test_steps.py]: So those are variable that are passed between steps?
If so how do you clean them up?
Aren't you suppose to use the `world` variable?

File: PythonRegression/tests/features/steps/api_test_steps.py [hunk 13]
```diff
@@ -13,52 +14,98 @@
...
+
+    if 'threads' not in config:
+        config['threads'] = {}
+    config['threads'][apiCall] = new_thread
+    #Wait 3 seconds to give node time to respond
+    sleep(3)
+
+
+@step(r'all api call threads are joined')
+def join_threads(step):
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/api_test_steps.py]: very non behavioral step...

File: PythonRegression/tests/features/steps/api_test_steps.py [hunk 14]
```diff
@@ -13,52 +14,98 @@
...
+        logger.info('Joining current thread: {}'.format(current_thread.name))
+        current_thread.join(5)
+        if current_thread.is_alive():
+            logger.info('Thread is still alive'.format(current_thread.name))
+        else:
+            logger.info("Thread is dead")
+
+
+@step(r'while checking for running threads, there should be "(\d+)" running')
+def check_thread_count(step,numThreads):
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/api_test_steps.py]: very non behavioral step...

File: PythonRegression/tests/features/steps/api_test_steps.py [hunk 15]
```diff
@@ -288,7 +336,7 @@ def fetch_response(apiCall):
 
 
 def check_neighbors(step):
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/api_test_steps.py]: Maybe add a docstring here?
Not sure what this function does

File: PythonRegression/util/static_vals.py [hunk 16]
```diff
@@ -4,28 +4,59 @@
...
 TEST_NEIGHBORS = [u'udp://178.128.236.6:14600',u'udp://167.99.178.3:14600'] 
 
-TEST_HASH = ["KEHXTIVIBJHIQOWXQ9BSDMIDXLNO9WJQUGOSKVTMFMAHNSAYDSFWGOYQWINLCAMBASPUATDNVCFT99999"]
+TEST_HASH = "NMPXODIWSCTMWRTQ9AI9VROYNFACWCQDXDSJLNC9HKCECBCGQTBUBKVROXZLQWAZRQUGIJTLTMAMSH999"
+
+TEST_TIP_LIST = ["SBKWTQWCFTF9DBZHJKQJKU9LXMZD9BMWJIJLZCCZYJFWIBGYYQBJOWWFWIHDEDTIHUB9PMOWZVCPKV999"]
 
 TEST_ADDRESS = "TEST9TRANSACTION9TEST9TRANSACTION9TEST9TRANSACTION9TEST9TRANSACTION9TEST999999999"
+TEST_EMPTY_ADDRESS = "EMPTY9BALANCE9TEST999999999999999999999999999999999999999999999999999999999999999"
+TEST_STORE_ADDRESS = "STORE9AND9FIND9999999999999999999999999999999999999999999999999999999999999999999"
```
Comment (GalRogozinski) [inline PythonRegression/util/static_vals.py]: A cool thing I found:
http://code.activestate.com/recipes/65207-constants-in-python/?in=user-97991

However, the zen of python implies an attitude that one should know what he is doing and don't need extra protections, so you don't have to use the solution in the link (just wanted to share).

Anyhows even when you declare constants in step files make them upper case to make sure no one changes them.

File: PythonRegression/tests/features/machine1/1_api_tests.feature [hunk 17]
```diff
@@ -38,56 +53,174 @@ Feature: Test API calls on Machine 1
...
+        |trunk_transaction      |TEST_HASH                      |staticValue    |
+        |branch_transaction     |TEST_HASH                      |staticValue    |
+        |min_weight_magnitude   |50                             |int            |
+
+        And "interruptAttachingToTangle" is called on one thread in "nodeA" with:
+        |keys                   |values                         |type           |
+
+        When all api call threads are joined
+        #The only active thread should be the main testing thread
+        Then while checking for running threads, there should be "1" running
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/machine1/1_api_tests.feature]: I doubt that this actually tests whether `interruptAttachingToTangle` really works...
I think you can put this code in a git stash (or save it somehow).

You can open a separate issue for it.

Comment (DyrellC) [reply PythonRegression/tests/features/machine1/1_api_tests.feature]: The test now ensures that the interrupt causes `attachToTangle` to return an empty tryte list

File: PythonRegression/tests/features/steps/api_test_steps.py (hunk 14 above)
Comment (DyrellC) [reply PythonRegression/tests/features/steps/api_test_steps.py]: Step has been replaced

File: PythonRegression/tests/features/steps/api_test_steps.py (hunk 13 above)
Comment (DyrellC) [reply PythonRegression/tests/features/steps/api_test_steps.py]: Step has been removed

File: PythonRegression/tests/features/steps/api_test_steps.py (hunk 12 above)
Comment (DyrellC) [reply PythonRegression/tests/features/steps/api_test_steps.py]: This will be done in a different PR

File: PythonRegression/tests/features/steps/api_test_steps.py (hunk 11 above)
Comment (DyrellC) [reply PythonRegression/tests/features/steps/api_test_steps.py]: This shouldn't have been left in the file and has been since removed

File: PythonRegression/tests/features/machine1/1_api_tests.feature [hunk 18]
```diff
@@ -1,93 +1,228 @@
...
+	#Values can be found in util/static_vals.py
+	Scenario: GetInclusionStates is called
+		Given "getInclusionStates" is called on "nodeA" with:
+		|keys           |values				|type               |
+		|transactions   |TEST_HASH			|staticList         |
+		|tips           |TEST_TIP_LIST			|staticValue        |
+
+		Then the response for "getInclusionStates" should return with:
+		|keys			|values			|type               |
+		|states			|[False]		|string             |
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/machine1/1_api_tests.feature]: It isn't intuitive that you have in the value a list and the type is string

File: PythonRegression/tests/features/machine1/1_api_tests.feature [hunk 19]
```diff
@@ -1,93 +1,228 @@
...
+	#Address can be found in util/static_vals.py
+	Scenario: GetBalances is called
+		Given "getBalances" is called on "nodeA" with:
+		|keys           |values			        |type               |
+		|addresses      |TEST_EMPTY_ADDRESS	        |staticList         |
+		|threshold      |100			        |int                |
+
+		Then the response for "getBalances" should return with:
+		|keys           |values                         |type		    |
+		|balances       |[0]			        |string     	    |
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/machine1/1_api_tests.feature]: Why isn't the type int?

File: PythonRegression/tests/features/steps/api_test_steps.py [hunk 20]
```diff
@@ -1,64 +1,118 @@
...
+    response = api_utils.fetch_call(apiCall, api, options)
+
+    assert type(response) is dict, 'There may be something wrong with the response format: {}'.format(response)
     
     responses[apiCall] = {}
     responses[apiCall][nodeName] = response
 
+#This method is identical to the method above, but creates a new thread
+#TODO: Find a way to pass aloe.world variable between threads to eliminate need for duplicate code
+@step(r'"([^"]*)" is called on one thread in "([^"]*)" with:')
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/api_test_steps.py]: Instead of writing "thread" write "parallel". Try to keep steps behavioral.

File: PythonRegression/tests/features/steps/api_test_steps.py [hunk 21]
```diff
@@ -1,64 +1,118 @@
...
+    new_thread.start()
+
+    if 'threads' not in config:
+        config['threads'] = {}
+    config['threads'][apiCall] = new_thread
+    #Wait 3 seconds to give node time to respond
+    sleep(3)
+
+
+@step(r'the "([^"]*)" thread should return with:')
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/api_test_steps.py]: Instead of writing "thread" write "parallel". Try to keep steps behavioral.

File: PythonRegression/tests/features/steps/api_test_steps.py [hunk 22]
```diff
@@ -107,14 +214,38 @@ def compare_response(step):
                     assert str(responseKeys[responseKeyVal]) == str(keys[responseKeyVal])
             except:
                 logger.debug("No values to verify response with")        
- 
+
+ ###
+ #Create Inconsistent transaction
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/api_test_steps.py]: Can you add an explanation to the docs as to why the tx is inconsistent?

File: PythonRegression/tests/features/steps/api_test_steps.py [hunk 23]
```diff
@@ -128,53 +259,20 @@ def check_trytes(step,trytes):
...
 @step(r'"([^"]*)" and "([^"]*)" are neighbors')
 def make_neighbors(step,node1,node2):
-    host1 = world.machine['nodes'][node1]['host']
-    port1 = world.machine['nodes'][node1]['ports']['gossip-udp']
-    host2 = world.machine['nodes'][node2]['host']
-    port2 = world.machine['nodes'][node2]['ports']['gossip-udp']
+    host1 = world.machine['nodes'][node1]['podip']
+    port1 = world.machine['nodes'][node1]['clusterip_ports']['gossip-udp']
+    host2 = world.machine['nodes'][node2]['podip']
+    port2 = world.machine['nodes'][node2]['clusterip_ports']['gossip-udp']
```
Comment (GalRogozinski) [inline PythonRegression/tests/features/steps/api_test_steps.py]: just wondering why  "ip" doesn't deserve "_" before?

File: PythonRegression/tests/features/steps/api_test_steps.py (hunk 15 above)
Comment (GalRogozinski) [reply PythonRegression/tests/features/steps/api_test_steps.py]: Ok,
I am fine with the implementation of this function for now :-)
Just remember that if you will ever need to change more than 2 neighbors change this to be dynamic and don't write another function.

File: PythonRegression/tests/features/steps/api_test_steps.py (hunk 23 above)
Comment (DyrellC) [reply PythonRegression/tests/features/steps/api_test_steps.py]: @karimodm
